src/components/cuda_pw/linux-cuda.c

```
@@ -0,0 +1,377 @@
+    .ntv_code_to_descr = cuda_ntv_code_to_descr,
+
+    .init_control_state = cuda_init_control_state,
+    .set_domain = cuda_set_domain,
+    .update_control_state = cuda_update_control_state,
+    .cleanup_eventset = cuda_cleanup_eventset,
+
+    .start = cuda_start,
+    .stop = cuda_stop,
+    .read = cuda_read,
+    .reset = cuda_reset,
+};
+
+unsigned _cuda_pw_lock;
+
+static int cuda_init_component(int cidx)
+{
+    COMPDBG("Entering with component idx: %d\n", cidx);
+
+    _cuda_pw_vector.cmp_info.CmpIdx = cidx;
+    _cuda_pw_vector.cmp_info.num_native_events = -1;
+    _cuda_pw_vector.cmp_info.num_cntrs = -1;
+    _cuda_pw_lock = PAPI_NUM_LOCK + NUM_INNER_LOCK + cidx;
+
+    _cuda_pw_vector.cmp_info.initialized = 1;
+    _cuda_pw_vector.cmp_info.disabled = PAPI_EDELAY_INIT;
+    sprintf(_cuda_pw_vector.cmp_info.disabled_reason,
+        "Not initialized. Access component events to initialize it.");
+    return PAPI_EDELAY_INIT;
+}
+
+static int cuda_shutdown_component(void)
+{
+    COMPDBG("Entering.\n");
+    int res = PAPI_OK;
+    free_event_name_list(&global_event_names);  // Free the global event names added dynamically
+
+    if (!_cuda_pw_vector.cmp_info.initialized ||
+    _cuda_pw_vector.cmp_info.disabled != PAPI_OK)
+    return PAPI_OK;
+
+    _cuda_pw_vector.cmp_info.initialized = 0;
+
+    // res = cupti_shutdown();
+    // Reverse of private init.
+    return res;
+}
+
+static int cuda_init_private(void)
+{
+    int res = PAPI_OK;
+    const char *disabled_reason;
+    COMPDBG("Entering.\n");
+    res = cupti_init(&disabled_reason);
+    if (res != PAPI_OK) {
+        goto fn_exit;
+    }
+
+    // Initialize global_event_names array
+    res = initialize_dynamic_event_list(&global_event_names);
+    if (res != PAPI_OK)
+        return res;
+
+    _cuda_pw_vector.cmp_info.disabled = PAPI_OK;
+    strcpy(_cuda_pw_vector.cmp_info.disabled_reason, "");
+
+fn_exit:
+    return res;
+}
+
+static void check_n_initialize(void)
```

This should return an error code.

---

Agree.

---

done.